File: orion-platform-service/src/services/self-healing/SelfHealingRepository.ts
import { DatabasePool } from '../database';
/**
 * SelfHealingRepository - Database layer for Self-Healing operations
 */

import {
  HealingStrategy,
  HealingAction,
  IncidentType,
  IncidentSeverity,
  IncidentStatus,
  HealingResult,
  ApprovalRequest,
  RiskLevel,
} from './types';

export interface SelfHealingRule {
  id: string;
  tenant_id: string;
  name: string;
  trigger_condition: Record<string, any>;
  action: Record<string, any>;
  enabled: boolean;
  execution_count: number;
  last_executed: Date | null;
  created_at: Date;
  updated_at: Date;
}

export interface SelfHealingExecution {
  id: string;
  rule_id: string;
  trigger_event: Record<string, any>;
  status: string;
  result: Record<string, any> | null;
  error_message: string | null;
  started_at: Date;
  completed_at: Date | null;
}

/**
 * Database representation of a healing incident
 */
export interface HealingIncidentRow {
  id: string;
  alert_id: string | null;
  type: string;
  severity: string;
  app_name: string;
  environment: string;
  strategy_id: string | null;
  strategy_name: string | null;
  actions: HealingAction[];
  status: string;
  attempts: number;
  approval_status: string | null;
  approval_request_id: string | null;
  result: HealingResult | null;
  error: string | null;
  tags: Record<string, string> | null;
  started_at: Date;
  completed_at: Date | null;
}

/**
 * Database representation of an approval request
 */
export interface ApprovalRequestRow {
  id: string;
  incident_id: string;
  title: string;
  description: string | null;
  risk_level: string;
  recommended_actions: HealingAction[];
  status: string;
  requested_by: string;
  approved_by: string | null;
  approval_reason: string | null;
  requested_at: Date;
  responded_at: Date | null;
  expires_at: Date | null;
}

export interface Ticket {
  id: string;
  tenant_id: string;
  title: string;
  description: string | null;
  type: string;
  priority: string;
  status: string;
  assignee_id: string | null;
  reporter_id: string | null;
  source: string | null;
  source_id: string | null;
  tags: string[];
  resolved_at: Date | null;
  created_at: Date;
  updated_at: Date;
}

export interface TicketComment {
  id: string;
  ticket_id: string;
  author_id: string | null;
  content: string;
  is_internal: boolean;
  created_at: Date;
}

export class SelfHealingRepository {
  constructor(private pool: DatabasePool) {}

  // ==================== Rules ====================

  async findRuleById(id: string): Promise<SelfHealingRule | null> {
    const result = await this.pool.query('SELECT * FROM self_healing_rules WHERE id = $1', [id]);
    return result.rows[0] || null;
  }

  async findAllRules(tenantId?: string): Promise<SelfHealingRule[]> {
    let query = 'SELECT * FROM self_healing_rules';
    const params: any[] = [];
    
    if (tenantId) { params.push(tenantId); query += ' WHERE tenant_id = $1'; }
    query += ' ORDER BY created_at DESC';
    
    const result = await this.pool.query(query, params);
    return result.rows;
  }

  async createRule(tenantId: string, name: string, triggerCondition: Record<string, any>, action: Record<string, any>): Promise<SelfHealingRule> {
    const result = await this.pool.query(
      `INSERT INTO self_healing_rules (tenant_id, name, trigger_condition, action, enabled)
       VALUES ($1, $2, $3, $4, true)
       RETURNING *`,
      [tenantId, name, triggerCondition, action]
    );
    return result.rows[0];
  }

  async updateRule(id: string, input: { name?: string; trigger_condition?: Record<string, any>; action?: Record<string, any>; enabled?: boolean }): Promise<SelfHealingRule | null> {
    const updates: string[] = [];
    const params: any[] = [];
    let paramIndex = 1;

    if (input.name !== undefined) { params.push(input.name); updates.push(`name = $${paramIndex++}`); }
    if (input.trigger_condition !== undefined) { params.push(JSON.stringify(input.trigger_condition)); updates.push(`trigger_condition = $${paramIndex++}`); }
    if (input.action !== undefined) { params.push(JSON.stringify(input.action)); updates.push(`action = $${paramIndex++}`); }
    if (input.enabled !== undefined) { params.push(input.enabled); updates.push(`enabled = $${paramIndex++}`); }

    if (updates.length === 0) return this.findRuleById(id);

    params.push(id);
    const result = await this.pool.query(
      `UPDATE self_healing_rules SET ${updates.join(', ')}, updated_at = NOW() WHERE id = $${paramIndex} RETURNING *`,
      params
    );
    return result.rows[0] || null;
  }

  async deleteRule(id: string): Promise<boolean> {
    const result = await this.pool.query('DELETE FROM self_healing_rules WHERE id = $1', [id]);
    return result.rowCount > 0;
  }

  async incrementExecutionCount(id: string): Promise<void> {
    await this.pool.query(
      'UPDATE self_healing_rules SET execution_count = execution_count + 1, last_executed = NOW() WHERE id = $1',
      [id]
    );
  }

  // ==================== Executions ====================

  async createExecution(ruleId: string, triggerEvent: Record<string, any>): Promise<SelfHealingExecution> {
    const result = await this.pool.query(
      `INSERT INTO self_healing_executions (rule_id, trigger_event, status)
       VALUES ($1, $2, 'running')
       RETURNING *`,
      [ruleId, triggerEvent]
    );
    return result.rows[0];
  }

  async completeExecution(id: string, status: string, result?: Record<string, any>, errorMessage?: string): Promise<SelfHealingExecution | null> {
    const updateResult = await this.pool.query(
      `UPDATE self_healing_executions SET status = $1, result = $2, error_message = $3, completed_at = NOW()
       WHERE id = $4 RETURNING *`,
      [status, result || null, errorMessage || null, id]
    );
    return updateResult.rows[0] || null;
  }

  async findExecutions(ruleId: string, limit: number = 10): Promise<SelfHealingExecution[]> {
    const result = await this.pool.query(
      'SELECT * FROM self_healing_executions WHERE rule_id = $1 ORDER BY started_at DESC LIMIT $2',
      [ruleId, limit]
    );
    return result.rows;
  }

  // ==================== Incidents ====================

  async createIncident(incident: {
    alert_id?: string;
    type: string;
    severity: string;
    app_name: string;
    environment: string;
    strategy_id?: string;
    strategy_name?: string;
    actions: HealingAction[];
    status: string;
    attempts?: number;
    approval_status?: string;
    approval_request_id?: string;
    tags?: Record<string, string>;
  }): Promise<HealingIncidentRow> {
    const result = await this.pool.query(
      `INSERT INTO self_healing_incidents
       (alert_id, type, severity, app_name, environment, strategy_id, strategy_name, actions, status, attempts, approval_status, approval_request_id, tags)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13)
       RETURNING *`,
      [
        incident.alert_id || null,
        incident.type,
        incident.severity,
        incident.app_name,
        incident.environment,
        incident.strategy_id || null,
        incident.strategy_name || null,
        incident.actions,
        incident.status,
        incident.attempts ?? 0,
        incident.approval_status || null,
        incident.approval_request_id || null,
        incident.tags || null,
      ]
    );
    return result.rows[0];
  }

  async findIncidentById(id: string): Promise<HealingIncidentRow | null> {
    const result = await this.pool.query(
      'SELECT * FROM self_healing_incidents WHERE id = $1',
      [id]
    );
    return result.rows[0] || null;
  }

  async findIncidents(filters: {
    appName?: string;
    environment?: string;
    type?: string;
    status?: string;
    severity?: string;
    startDate?: Date;
    endDate?: Date;
    limit?: number;
    offset?: number;
  }): Promise<{ rows: HealingIncidentRow[]; total: number }> {
    const conditions: string[] = [];
    const params: any[] = [];
    let paramIndex = 1;

    if (filters.appName) { params.push(filters.appName); conditions.push(`app_name = $${paramIndex++}`); }
    if (filters.environment) { params.push(filters.environment); conditions.push(`environment = $${paramIndex++}`); }
    if (filters.type) { params.push(filters.type); conditions.push(`type = $${paramIndex++}`); }
    if (filters.status) { params.push(filters.status); conditions.push(`status = $${paramIndex++}`); }
    if (filters.severity) { params.push(filters.severity); conditions.push(`severity = $${paramIndex++}`); }
    if (filters.startDate) { params.push(filters.startDate); conditions.push(`started_at >= $${paramIndex++}`); }
    if (filters.endDate) { params.push(filters.endDate); conditions.push(`started_at <= $${paramIndex++}`); }

    const whereClause = conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : '';

    // Count query
    const countResult = await this.pool.query(
      `SELECT COUNT(*) FROM self_healing_incidents ${whereClause}`,
      params
    );
    const total = parseInt(countResult.rows[0].count, 10);

    // Data query with pagination
    const limit = filters.limit ?? 50;
    const offset = filters.offset ?? 0;
    params.push(limit, offset);
    const dataResult = await this.pool.query(
      `SELECT * FROM self_healing_incidents ${whereClause}
       ORDER BY started_at DESC
       LIMIT $${paramIndex} OFFSET $${paramIndex + 1}`,
      params
    );

    return { rows: dataResult.rows, total };
  }

  async updateIncident(id: string, updates: {
    status?: string;
    strategy_id?: string;
    strategy_name?: string;
    actions?: HealingAction[];
    attempts?: number;
    approval_status?: string;
    approval_request_id?: string;
    result?: HealingResult;
    error?: string;
    completed_at?: Date;
  }): Promise<HealingIncidentRow | null> {
    const setClauses: string[] = [];
    const params: any[] = [];
    let paramIndex = 1;

    if (updates.status !== undefined) { params.push(updates.status); setClauses.push(`status = $${paramIndex++}`); }
    if (updates.strategy_id !== undefined) { params.push(updates.strategy_id); setClauses.push(`strategy_id = $${paramIndex++}`); }
    if (updates.strategy_name !== undefined) { params.push(updates.strategy_name); setClauses.push(`strategy_name = $${paramIndex++}`); }
    if (updates.actions !== undefined) { params.push(JSON.stringify(updates.actions)); setClauses.push(`actions = $${paramIndex++}`); }
    if (updates.attempts !== undefined) { params.push(updates.attempts); setClauses.push(`attempts = $${paramIndex++}`); }
    if (updates.approval_status !== undefined) { params.push(updates.approval_status); setClauses.push(`approval_status = $${paramIndex++}`); }
    if (updates.approval_request_id !== undefined) { params.push(updates.approval_request_id); setClauses.push(`approval_request_id = $${paramIndex++}`); }
    if (updates.result !== undefined) { params.push(JSON.stringify(updates.result)); setClauses.push(`result = $${paramIndex++}`); }
    if (updates.error !== undefined) { params.push(updates.error); setClauses.push(`error = $${paramIndex++}`); }
    if (updates.completed_at !== undefined) { params.push(updates.completed_at); setClauses.push(`completed_at = $${paramIndex++}`); }

    if (setClauses.length === 0) return this.findIncidentById(id);

    params.push(id);
    const result = await this.pool.query(
      `UPDATE self_healing_incidents SET ${setClauses.join(', ')} WHERE id = $${paramIndex} RETURNING *`,
      params
    );
    return result.rows[0] || null;
  }

  // ==================== Approvals ====================

  async createApprovalRequest(approval: {
    incident_id: string;
    title: string;
    description?: string;
    risk_level: string;
    recommended_actions: HealingAction[];
    status?: string;
    requested_by?: string;
    expires_at?: Date;
  }): Promise<ApprovalRequestRow> {
    const result = await this.pool.query(
      `INSERT INTO self_healing_approvals
       (incident_id, title, description, risk_level, recommended_actions, status, requested_by, expires_at)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
       RETURNING *`,
      [
        approval.incident_id,
        approval.title,
        approval.description || null,
        approval.risk_level,
        approval.recommended_actions,
        approval.status || 'pending',
        approval.requested_by || 'system',
        approval.expires_at || null,
      ]
    );
    return result.rows[0];
  }

  async findApprovalById(id: string): Promise<ApprovalRequestRow | null> {
    const result = await this.pool.query(
      'SELECT * FROM self_healing_approvals WHERE id = $1',
      [id]
    );
    return result.rows[0] || null;
  }

  async findApprovalsByStatus(status?: string): Promise<ApprovalRequestRow[]> {
    if (status) {
      const result = await this.pool.query(
        'SELECT * FROM self_healing_approvals WHERE status = $1 ORDER BY requested_at DESC',
        [status]
      );
      return result.rows;
    }
    const result = await this.pool.query(
      'SELECT * FROM self_healing_approvals ORDER BY requested_at DESC'
    );
    return result.rows;
  }

  async updateApprovalRequest(id: string, updates: {
    status?: string;
    approved_by?: string;
    approval_reason?: string;
    responded_at?: Date;
  }): Promise<ApprovalRequestRow | null> {
    const setClauses: string[] = [];
    const params: any[] = [];
    let paramIndex = 1;

    if (updates.status !== undefined) { params.push(updates.status); setClauses.push(`status = $${paramIndex++}`); }
    if (updates.approved_by !== undefined) { params.push(updates.approved_by); setClauses.push(`approved_by = $${paramIndex++}`); }
    if (updates.approval_reason !== undefined) { params.push(updates.approval_reason); setClauses.push(`approval_reason = $${paramIndex++}`); }
    if (updates.responded_at !== undefined) { params.push(updates.responded_at); setClauses.push(`responded_at = $${paramIndex++}`); }

    if (setClauses.length === 0) return this.findApprovalById(id);

    params.push(id);
    const result = await this.pool.query(
      `UPDATE self_healing_approvals SET ${setClauses.join(', ')} WHERE id = $${paramIndex} RETURNING *`,
      params
    );
    return result.rows[0] || null;
  }

  async markExpiredApprovals(): Promise<number> {
    const result = await this.pool.query(
      `UPDATE self_healing_approvals
       SET status = 'expired'
       WHERE status = 'pending' AND expires_at IS NOT NULL AND expires_at < NOW()`
    );
    return result.rowCount ?? 0;
  }
}

File: orion-selfhealing-svc-go/internal/repository/selfhealing_repository.go
package repository

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"orion/selfhealing-svc-go/internal/models"

	"github.com/jmoiron/sqlx"
)

// Repository provides data access for all self-healing entities.
type Repository struct {
	db *sqlx.DB
}

// NewRepository creates a new Repository.
func NewRepository(db *sqlx.DB) *Repository {
	return &Repository{db: db}
}

// ==================== Healing Rules ====================

// Create inserts a new healing rule.
func (r *Repository) Create(ctx context.Context, d *models.HealingRule) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO healing_rules (id, tenant_id, name, trigger_type, action, status, config, enabled, execution_count)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)`,
		d.ID, d.TenantID, d.Name, d.TriggerType, d.Action, d.Status, d.Config, d.Enabled, d.ExecutionCount)
	return err
}

// List returns healing rules with pagination.
func (r *Repository) List(ctx context.Context, tenantID string, offset, limit int) ([]models.HealingRule, error) {
	var items []models.HealingRule
	err := r.db.SelectContext(ctx, &items,
		`SELECT * FROM healing_rules WHERE tenant_id=$1 ORDER BY created_at DESC OFFSET $2 LIMIT $3`,
		tenantID, offset, limit)
	return items, err
}

// GetByID returns a healing rule by ID.
func (r *Repository) GetByID(ctx context.Context, tenantID, id string) (*models.HealingRule, error) {
	var d models.HealingRule
	err := r.db.GetContext(ctx, &d,
		`SELECT * FROM healing_rules WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	if err != nil {
		return nil, err
	}
	return &d, nil
}

// Update updates a healing rule.
func (r *Repository) Update(ctx context.Context, tenantID, id string, req *models.UpdateHealingRuleRequest) (*models.HealingRule, error) {
	setClauses := []string{}
	args := []interface{}{}
	idx := 1

	if req.Name != nil {
		setClauses = append(setClauses, fmt.Sprintf("name=$%d", idx))
		args = append(args, *req.Name)
		idx++
	}
	if req.TriggerType != nil {
		setClauses = append(setClauses, fmt.Sprintf("trigger_type=$%d", idx))
		args = append(args, *req.TriggerType)
		idx++
	}
	if req.Action != nil {
		setClauses = append(setClauses, fmt.Sprintf("action=$%d", idx))
		args = append(args, *req.Action)
		idx++
	}
	if req.Config != nil {
		setClauses = append(setClauses, fmt.Sprintf("config=$%d", idx))
		args = append(args, req.Config)
		idx++
	}
	if req.Enabled != nil {
		setClauses = append(setClauses, fmt.Sprintf("enabled=$%d", idx))
		args = append(args, *req.Enabled)
		idx++
	}

	if len(setClauses) == 0 {
		return r.GetByID(ctx, tenantID, id)
	}

	setClauses = append(setClauses, "updated_at=NOW()")
	args = append(args, id, tenantID)

	query := fmt.Sprintf("UPDATE healing_rules SET %s WHERE id=$%d AND tenant_id=$%d RETURNING *",
		strings.Join(setClauses, ", "), idx, idx+1)

	var d models.HealingRule
	err := r.db.GetContext(ctx, &d, query, args...)
	if err != nil {
		return nil, err
	}
	return &d, nil
}

// Delete removes a healing rule.
func (r *Repository) Delete(ctx context.Context, tenantID, id string) error {
	_, err := r.db.ExecContext(ctx,
		`DELETE FROM healing_rules WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	return err
}

// Count returns the total number of healing rules for a tenant.
func (r *Repository) Count(ctx context.Context, tenantID string) (int, error) {
	var count int
	err := r.db.GetContext(ctx, &count,
		`SELECT COUNT(*) FROM healing_rules WHERE tenant_id=$1`, tenantID)
	return count, err
}

// IncrementExecutionCount increments the execution count and updates last_triggered.
func (r *Repository) IncrementExecutionCount(ctx context.Context, id string) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE healing_rules SET execution_count = execution_count + 1, last_triggered = NOW(), updated_at = NOW() WHERE id = $1`, id)
	return err
}

// ==================== Healing Strategies ====================

// CreateStrategy inserts a new healing strategy.
func (r *Repository) CreateStrategy(ctx context.Context, s *models.HealingStrategy) error {
	actions, _ := json.Marshal(s.Actions)
	conditions, _ := json.Marshal(s.Conditions)
	envs, _ := json.Marshal(s.Environments)

	_, err := r.db.ExecContext(ctx,
		`INSERT INTO healing_strategies (id, name, trigger_type, actions, conditions, confidence, enabled,
		  description, environments, max_retries, retry_cooldown_ms)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
		 ON CONFLICT (id) DO UPDATE SET
		 	name=$2, trigger_type=$3, actions=$4, conditions=$5, confidence=$6, enabled=$7,
		 	description=$8, environments=$9, max_retries=$10, retry_cooldown_ms=$11, updated_at=NOW()`,
		s.ID, s.Name, string(s.TriggerType), actions, conditions, s.Confidence, s.Enabled,
		s.Description, envs, &s.MaxRetries, &s.RetryCooldownMs)
	return err
}

// FindStrategyByID returns a strategy by ID.
func (r *Repository) FindStrategyByID(ctx context.Context, id string) (*models.HealingStrategy, error) {
	var s models.HealingStrategy
	err := r.db.GetContext(ctx, &s,
		`SELECT id, name, trigger_type, actions, conditions, confidence, enabled,
		  description, environments, max_retries, retry_cooldown_ms, created_at, updated_at
		 FROM healing_strategies WHERE id=$1`, id)
	if err != nil {
		return nil, err
	}
	return &s, nil
}

// FindEnabledStrategies returns all enabled strategies.
func (r *Repository) FindEnabledStrategies(ctx context.Context) ([]models.HealingStrategy, error) {
	var items []models.HealingStrategy
	err := r.db.SelectContext(ctx, &items,
		`SELECT id, name, trigger_type, actions, conditions, confidence, enabled,
		  description, environments, max_retries, retry_cooldown_ms, created_at, updated_at
		 FROM healing_strategies WHERE enabled=$1 ORDER BY confidence DESC`, true)
	return items, err
}

// FindAllStrategies returns all strategies.
func (r *Repository) FindAllStrategies(ctx context.Context) ([]models.HealingStrategy, error) {
	var items []models.HealingStrategy
	err := r.db.SelectContext(ctx, &items,
		`SELECT id, name, trigger_type, actions, conditions, confidence, enabled,
		  description, environments, max_retries, retry_cooldown_ms, created_at, updated_at
		 FROM healing_strategies ORDER BY created_at DESC`)
	return items, err
}

// DeleteStrategy removes a strategy.
func (r *Repository) DeleteStrategy(ctx context.Context, id string) error {
	_, err := r.db.ExecContext(ctx, `DELETE FROM healing_strategies WHERE id=$1`, id)
	return err
}

// SetStrategyEnabled updates the enabled flag of a strategy.
func (r *Repository) SetStrategyEnabled(ctx context.Context, id string, enabled bool) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE healing_strategies SET enabled=$1, updated_at=NOW() WHERE id=$2`, enabled, id)
	return err
}

// SeedStrategy inserts a strategy only if it does not exist.
func (r *Repository) SeedStrategy(ctx context.Context, s *models.HealingStrategy) error {
	actions, _ := json.Marshal(s.Actions)
	conditions, _ := json.Marshal(s.Conditions)
	envs, _ := json.Marshal(s.Environments)

	_, err := r.db.ExecContext(ctx,
		`INSERT INTO healing_strategies (id, name, trigger_type, actions, conditions, confidence, enabled,
		  description, environments, max_retries, retry_cooldown_ms)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
		 ON CONFLICT (id) DO NOTHING`,
		s.ID, s.Name, string(s.TriggerType), actions, conditions, s.Confidence, s.Enabled,
		s.Description, envs, &s.MaxRetries, &s.RetryCooldownMs)
	return err
}

// ==================== Healing Incidents ====================

// CreateIncident inserts a new healing incident.
func (r *Repository) CreateIncident(ctx context.Context, inc *models.HealingIncident) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO healing_incidents
		 (id, tenant_id, alert_id, type, severity, app_name, environment,
		  strategy_id, strategy_name, actions, status, attempts,
		  approval_status, approval_request_id, tags)
		 VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15)`,
		inc.ID, inc.TenantID, inc.AlertID, string(inc.Type), string(inc.Severity),
		inc.AppName, inc.Environment, inc.StrategyID, inc.StrategyName,
		inc.Actions, string(inc.Status), inc.Attempts,
		nullApprovalStr(inc.ApprovalStatus), inc.ApprovalRequestID, inc.Tags)
	return err
}

// FindIncidentByID returns an incident by ID.
func (r *Repository) FindIncidentByID(ctx context.Context, id string) (*models.HealingIncident, error) {
	var db models.HealingIncidentDB
	err := r.db.GetContext(ctx, &db,
		`SELECT id, tenant_id, alert_id, type, severity, app_name, environment,
		        strategy_id, strategy_name, actions, status, attempts,
		        approval_status, approval_request_id, result, error, tags, started_at, completed_at
		 FROM healing_incidents WHERE id=$1`, id)
	if err != nil {
		return nil, err
	}
	inc := db.ToIncident()
	return &inc, nil
}

// FindIncidents returns incidents with filters and pagination.
func (r *Repository) FindIncidents(ctx context.Context, tenantID string, q *models.HistoryQuery) ([]models.HealingIncident, int, error) {
	conditions := []string{"tenant_id = $1"}
	args := []interface{}{tenantID}
	idx := 2

	if q.AppName != nil {
		conditions = append(conditions, fmt.Sprintf("app_name = $%d", idx))
		args = append(args, *q.AppName)
		idx++
	}
	if q.Environment != nil {
		conditions = append(conditions, fmt.Sprintf("environment = $%d", idx))
		args = append(args, *q.Environment)
		idx++
	}
	if q.Type != nil {
		conditions = append(conditions, fmt.Sprintf("type = $%d", idx))
		args = append(args, *q.Type)
		idx++
	}
	if q.Status != nil {
		conditions = append(conditions, fmt.Sprintf("status = $%d", idx))
		args = append(args, *q.Status)
		idx++
	}
	if q.Severity != nil {
		conditions = append(conditions, fmt.Sprintf("severity = $%d", idx))
		args = append(args, *q.Severity)
		idx++
	}
	if q.StartDate != nil {
		conditions = append(conditions, fmt.Sprintf("started_at >= $%d", idx))
		args = append(args, *q.StartDate)
		idx++
	}
	if q.EndDate != nil {
		conditions = append(conditions, fmt.Sprintf("started_at <= $%d", idx))
		args = append(args, *q.EndDate)
		idx++
	}

	where := strings.Join(conditions, " AND ")

	// Count
	var total int
	countQuery := fmt.Sprintf("SELECT COUNT(*) FROM healing_incidents WHERE %s", where)
	if err := r.db.GetContext(ctx, &total, countQuery, args...); err != nil {
		return nil, 0, err
	}

	// Data
	limit := q.PageSize
	if limit <= 0 {
		limit = 50
	}
	if limit > 100 {
		limit = 100
	}
	offset := (q.Page - 1) * limit
	if q.Page <= 0 {
		offset = 0
	}
	args = append(args, limit, offset)
	dataQuery := fmt.Sprintf(
		`SELECT id, tenant_id, alert_id, type, severity, app_name, environment,
		        strategy_id, strategy_name, actions, status, attempts,
		        approval_status, approval_request_id, result, error, tags, started_at, completed_at
		 FROM healing_incidents WHERE %s ORDER BY started_at DESC LIMIT $%d OFFSET $%d`,
		where, idx, idx+1)

	var dbRows []models.HealingIncidentDB
	if err := r.db.SelectContext(ctx, &dbRows, dataQuery, args...); err != nil {
		return nil, 0, err
	}

	incidents := make([]models.HealingIncident, len(dbRows))
	for i, row := range dbRows {
		incidents[i] = row.ToIncident()
	}
	return incidents, total, nil
}

// UpdateIncident updates an incident.
func (r *Repository) UpdateIncident(ctx context.Context, id string, updates map[string]interface{}) error {
	setClauses := []string{}
	args := []interface{}{}
	idx := 1

	for col, val := range updates {
		setClauses = append(setClauses, fmt.Sprintf("%s=$%d", col, idx))
		args = append(args, val)
		idx++
	}

	if len(setClauses) == 0 {
		return nil
	}

	args = append(args, id)
	query := fmt.Sprintf("UPDATE healing_incidents SET %s WHERE id=$%d", strings.Join(setClauses, ", "), idx)
	_, err := r.db.ExecContext(ctx, query, args...)
	return err
}

// GetAllIncidentsForStats returns all incidents matching filters for stats computation.
func (r *Repository) GetAllIncidentsForStats(ctx context.Context, tenantID string, q *models.EffectivenessQuery) ([]models.HealingIncident, error) {
	conditions := []string{"tenant_id = $1"}
	args := []interface{}{tenantID}
	idx := 2

	if q.AppName != nil {
		conditions = append(conditions, fmt.Sprintf("app_name = $%d", idx))
		args = append(args, *q.AppName)
		idx++
	}
	if q.Environment != nil {
		conditions = append(conditions, fmt.Sprintf("environment = $%d", idx))
		args = append(args, *q.Environment)
		idx++
	}
	if q.StartDate != nil {
		conditions = append(conditions, fmt.Sprintf("started_at >= $%d", idx))
		args = append(args, *q.StartDate)
		idx++
	}
	if q.EndDate != nil {
		conditions = append(conditions, fmt.Sprintf("started_at <= $%d", idx))
		args = append(args, *q.EndDate)
		idx++
	}

	where := strings.Join(conditions, " AND ")
	query := fmt.Sprintf(
		`SELECT id, tenant_id, alert_id, type, severity, app_name, environment,
		        strategy_id, strategy_name, actions, status, attempts,
		        approval_status, approval_request_id, result, error, tags, started_at, completed_at
		 FROM healing_incidents WHERE %s ORDER BY started_at DESC LIMIT 10000`, where)

	var dbRows []models.HealingIncidentDB
	if err := r.db.SelectContext(ctx, &dbRows, query, args...); err != nil {
		return nil, err
	}

	incidents := make([]models.HealingIncident, len(dbRows))
	for i, row := range dbRows {
		incidents[i] = row.ToIncident()
	}
	return incidents, nil
}

// ==================== Approval Requests ====================

// CreateApproval inserts a new approval request.
func (r *Repository) CreateApproval(ctx context.Context, a *models.ApprovalRequest) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO healing_approvals
		 (id, tenant_id, incident_id, title, description, risk_level, recommended_actions,
		  status, requested_by, expires_at)
		 VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)`,
		a.ID, a.TenantID, a.IncidentID, a.Title, a.Description, string(a.RiskLevel),
		a.RecommendedActions, string(a.Status), a.RequestedBy, a.ExpiresAt)
	return err
}

// FindApprovalByID returns an approval request by ID.
func (r *Repository) FindApprovalByID(ctx context.Context, id string) (*models.ApprovalRequest, error) {
	var a models.ApprovalRequest
	err := r.db.GetContext(ctx, &a,
		`SELECT id, tenant_id, incident_id, title, description, risk_level, recommended_actions,
		        status, requested_by, approved_by, approval_reason, requested_at, responded_at, expires_at
		 FROM healing_approvals WHERE id=$1`, id)
	if err != nil {
		return nil, err
	}
	return &a, nil
}

// FindApprovalsByStatus returns approval requests filtered by status.
func (r *Repository) FindApprovalsByStatus(ctx context.Context, tenantID string, status *string) ([]models.ApprovalRequest, error) {
	var items []models.ApprovalRequest
	var err error
	if status != nil {
		err = r.db.SelectContext(ctx, &items,
			`SELECT id, tenant_id, incident_id, title, description, risk_level, recommended_actions,
			        status, requested_by, approved_by, approval_reason, requested_at, responded_at, expires_at
			 FROM healing_approvals WHERE tenant_id=$1 AND status=$2 ORDER BY requested_at DESC`,
			tenantID, *status)
	} else {
		err = r.db.SelectContext(ctx, &items,
			`SELECT id, tenant_id, incident_id, title, description, risk_level, recommended_actions,
			        status, requested_by, approved_by, approval_reason, requested_at, responded_at, expires_at
			 FROM healing_approvals WHERE tenant_id=$1 ORDER BY requested_at DESC`,
			tenantID)
	}
	return items, err
}

// UpdateApproval updates an approval request.
func (r *Repository) UpdateApproval(ctx context.Context, id string, status string, approvedBy *string, reason *string) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE healing_approvals SET status=$1, approved_by=$2, approval_reason=$3, responded_at=NOW()
		 WHERE id=$4`,
		status, approvedBy, reason, id)
	return err
}

// MarkExpiredApprovals marks all pending expired approvals as expired.
func (r *Repository) MarkExpiredApprovals(ctx context.Context) (int64, error) {
	result, err := r.db.ExecContext(ctx,
		`UPDATE healing_approvals SET status='expired'
		 WHERE status='pending' AND expires_at IS NOT NULL AND expires_at < NOW()`)
	if err != nil {
		return 0, err
	}
	return result.RowsAffected()
}

// ==================== Rule Executions ====================

// CreateExecution inserts a new rule execution.
func (r *Repository) CreateExecution(ctx context.Context, e *models.HealingExecution) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO healing_executions (id, rule_id, trigger_event, status)
		 VALUES ($1, $2, $3, $4)`,
		e.ID, e.RuleID, e.TriggerEvent, string(e.Status))
	return err
}

// CompleteExecution updates an execution as completed.
func (r *Repository) CompleteExecution(ctx context.Context, id string, status models.ExecutionStatus, result models.JSONB, errMsg *string) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE healing_executions SET status=$1, result=$2, error_message=$3, completed_at=NOW()
		 WHERE id=$4`,
		string(status), result, errMsg, id)
	return err
}

// FindExecutionsByRule returns executions for a given rule.
func (r *Repository) FindExecutionsByRule(ctx context.Context, ruleID string, limit int) ([]models.HealingExecution, error) {
	var items []models.HealingExecution
	err := r.db.SelectContext(ctx, &items,
		`SELECT * FROM healing_executions WHERE rule_id=$1 ORDER BY started_at DESC LIMIT $2`,
		ruleID, limit)
	return items, err
}

// ==================== Guardian Audit ====================

// InsertAudit inserts a guardian audit entry.
func (r *Repository) InsertAudit(ctx context.Context, a *models.HealingAuditEntry) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO guardian_audit (id, tenant_id, incident_id, action_type, target, environment,
		  risk_level, approvers, executor, status, reason, result)
		 VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12)`,
		a.ID, "", a.IncidentID, a.ActionType, a.Target, a.Environment,
		string(a.RiskLevel), a.Approvers, a.Executor, a.Status, a.Reason, a.Result)
	return err
}

// FindAuditsByIncident returns audit entries for an incident.
func (r *Repository) FindAuditsByIncident(ctx context.Context, tenantID string, incidentID string, limit int) ([]models.HealingAuditEntry, error) {
	var items []models.HealingAuditEntry
	err := r.db.SelectContext(ctx, &items,
		`SELECT * FROM guardian_audit WHERE tenant_id=$1 AND incident_id=$2 ORDER BY created_at DESC LIMIT $3`,
		tenantID, incidentID, limit)
	return items, err
}

// CountAudits returns total audit count for a tenant.
func (r *Repository) CountAudits(ctx context.Context, tenantID string) (int, error) {
	var count int
	err := r.db.GetContext(ctx, &count,
		`SELECT COUNT(*) FROM guardian_audit WHERE tenant_id=$1`, tenantID)
	return count, err
}

// ==================== Helpers ====================

func nullApprovalStr(s *models.ApprovalStatus) sql.NullString {
	if s == nil {
		return sql.NullString{}
	}
	return sql.NullString{String: string(*s), Valid: true}
}

// timePtr returns a pointer to a time value.
func timePtr(t time.Time) *time.Time {
	return &t
}

// jsonMarshal is a helper to marshal to JSON bytes.
func jsonMarshal(v interface{}) []byte {
	b, _ := json.Marshal(v)
	return b
}
